pkg/kubectl/history.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_1"
+	"k8s.io/kubernetes/pkg/runtime"
+	deploymentutil "k8s.io/kubernetes/pkg/util/deployment"
+	"k8s.io/kubernetes/pkg/util/errors"
+)
+
+const (
+	ChangeCauseAnnotation = "kubernetes.io/change-cause"
+)
+
+// HistoryViewer provides an interface for resources that can be rolled back.
+type HistoryViewer interface {
+	History(namespace, name string) (HistoryInfo, error)
+}
+
+func HistoryViewerFor(kind unversioned.GroupKind, c clientset.Interface) (HistoryViewer, error) {
+	switch kind {
+	case extensions.Kind("Deployment"):
+		return &DeploymentHistoryViewer{c}, nil
+	}
+	return nil, fmt.Errorf("no history viewer has been implemented for %q", kind)
+}
+
+// HistoryInfo stores the mapping from revision to podTemplate and revision to change-cause
+type HistoryInfo struct {
+	RevisionToTemplate    map[int64]*api.PodTemplateSpec
+	RevisionToChangeCause map[int64]string
```

Comment:
PodTemplateSpec has ObjectMeta. How about storing the annotation in there and use only one map?